ArchiSteamFarm/Bot.cs

```diff
@@ -2836,6 +2840,290 @@ public sealed class Bot : IAsyncDisposable {
+
+					foreach (HashSet<uint>? result in results) {
+						if (result == null) {
+							return null;
+						}
+
+						firstPageResult.UnionWith(result);
+					}
+
+					return firstPageResult;
+			}
+		}
+
+		private async Task<HashSet<uint>?> GetPossiblyCompletedBadgeAppIDs(byte page) {
+			if (page == 0) {
+				throw new ArgumentNullException(nameof(page));
+			}
+
+			using IDocument? badgePage = await ArchiWebHandler.GetBadgePage(page).ConfigureAwait(false);
+
+			if (badgePage == null) {
+				ArchiLogger.LogGenericWarning(Strings.WarningCouldNotCheckBadges);
+
+				return null;
+			}
+
+			return GetPossiblyCompletedBadgeAppIDs(badgePage);
+		}
+
+		private HashSet<uint>? GetPossiblyCompletedBadgeAppIDs(IDocument badgePage) {
+			if (badgePage == null) {
+				throw new ArgumentNullException(nameof(badgePage));
+			}
+
+			List<IElement> linkElements = badgePage.SelectNodes("//a[@class='badge_craft_button']");
+
+			// We need to also select all badges that we have max level, as those will not display with a craft button
+			// Level 5 is maximum level for card badges according to https://steamcommunity.com/tradingcards/faq
+			linkElements.AddRange(badgePage.SelectNodes("//div[@class='badges_sheet']/div[contains(@class, 'badge_row') and .//div[@class='badge_info_description']/div[contains(text(), 'Level 5')]]/a[@class='badge_row_overlay']"));
+
+			if (linkElements.Count == 0) {
+				return new HashSet<uint>(0);
+			}
+
+			HashSet<uint> result = new HashSet<uint>(linkElements.Count);
+
+			foreach (string? badgeUri in linkElements.Select(htmlNode => htmlNode.GetAttribute("href"))) {
+				if (string.IsNullOrEmpty(badgeUri)) {
+					ArchiLogger.LogNullError(nameof(badgeUri));
+
+					return null;
+				}
+
+				// URIs to foil badges are the same as for normal badges except they end with "?border=1"
+				string? appIDText = badgeUri.Split('/', StringSplitOptions.RemoveEmptyEntries)[^1].Split('?', StringSplitOptions.RemoveEmptyEntries)[0];
+
+				if (!uint.TryParse(appIDText, out uint appID) || (appID == 0)) {
+					ArchiLogger.LogNullError(nameof(appID));
+
+					return null;
+				}
+
+				result.Add(appID);
+			}
+
+			return result;
+		}
+
+		private async Task SendCompletedSets() {
+			lock (SendCompleteTypesSemaphore) {
+				if (SendCompleteTypesScheduled) {
+					return;
+				}
+
+				SendCompleteTypesScheduled = true;
+			}
+
+			await SendCompleteTypesSemaphore.WaitAsync().ConfigureAwait(false);
+
+			try {
+				lock (SendCompleteTypesSemaphore) {
+					SendCompleteTypesScheduled = false;
+				}
+
+				HashSet<uint>? appIDs = await GetPossiblyCompletedBadgeAppIDs().ConfigureAwait(false);
+
+				if ((appIDs == null) || (appIDs.Count == 0)) {
+					return;
+				}
+
+				HashSet<Steam.Asset> inventory;
+
+				try {
+					inventory = await ArchiWebHandler.GetInventoryAsync(SteamID)
+						.Where(item => item.Tradable && appIDs.Contains(item.RealAppID) && BotConfig.CompleteTypesToSend.Contains(item.Type))
```

Comment:
We can avoid loading all the inventory if `BotConfig.CompleteTypesToSend` is empty.

Reply:
This method is not even called if `CompleteTypesToSend` is empty. If this method was public API I'd totally agree with you, but I don't think checking it twice is necessary in this case.

Reply:
That's true, it's a private method, and we already did a pre-check before. It's optional whether we check it twice here (and in general we shouldn't unless we can't be completely sure about all the calls - here we are pretty sure).

Reply:
Right, forgot about a check earlier, sorry.